Add logging module

neumayer opened the pull request.

Commit:
Add webconsole module.

- Opt in SecondBase module which starts a local web server.
- WebConsole looks for Widgets and hosts the ones it finds.
- SecondBase looks for WebConsole implementations and attempts to
start in only if it exists in classpath.
- WebConsole module will only start if the --enable-webconsole Flag
has been set.

Commit:
Add widget loading

Commit:
One less newline

Commit:
Throw SecondBaseException from example

Commit:
Add final for classes

neumayer committed:
Add final keyword and empty line (#19)

Commit:
Add json logging module

Commit:
Review fixes

Commit:
Add logging to example

Commit:
Rebase

Commit:
Rename and add tests

Commit:
Merge branch 'master' into logging

Commit:
Add annotation

acidmoose commented on jsonlogging/pom.xml:
```diff
@@ -0,0 +1,65 @@
+    <name>jsonlogging</name>
+
+    <properties>
+      <joda-time.version>2.3</joda-time.version>
```
Should either do versions in properties or in the dependencies. The mix makes it a bit cumbersome to get an overview.

neumayer replied:
Done. Although the idea was to put things that are used more than once in a property. I'm sure we'll now update all other pom files to follow the same pattern...

acidmoose replied:
joda-time is used once

neumayer replied:
The idea was that. Anyway, I agree it's easier to be consistent.

acidmoose commented on jsonlogging/src/main/java/org/secondbase/logging/JsonLogger.java:
```diff
@@ -0,0 +1,215 @@
+            final String service,
+            final String datacenter,
+            final String requestLoggerName,
+            final boolean json) {
```
So we have a JsonLogger class that can optionally _not_ use json? How about renaming the class to SecondBaseLogger?

neumayer replied:
Ok. There'll have to be some refactoring anyway after. Done.

Commit:
Remove environment parameter parsing

This will come at a future point.

acidmoose commented on example/src/main/java/org/secondbase/example/main/HelloJsonLogging.java:
```diff
@@ -0,0 +1,26 @@
+public final class HelloJsonLogging {
+
+    @Flag(name = "var")
+    private String variable = "";
```
Can remove the Flags when we're not demoing Flags specifically.

neumayer replied:
Yeah, I did. Thanks. Done. Done.

acidmoose commented on jsonlogging/src/main/java/org/secondbase/logging/JsonLoggerModule.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Flag(name = "datacenter",
+            description = "The datacenter the service runs in (eu-west-1|ap-southeast-1).")
+    private static String datacenter = "eu-west-1";
```
I don't think these should have defaults. At least not this, which is very specific to amazon.

neumayer replied:
I removed all defaults. This also means I can't abuse the loadOpts method any longer. One solution I see is a second loop over the modules and calling their init methods in that loop (after the parameters are parsed).

I'm not sure I'm convinced though. Right now I am reading the values in the main method only. Just to make this work for now.

acidmoose commented on jsonlogging/src/main/java/org/secondbase/logging/JsonLoggerModule.java:
```diff
@@ -0,0 +1,30 @@
+import org.secondbase.flags.Flag;
+
+public class JsonLoggerModule implements SecondBaseModule {
+
```
I want to move these to the (now deleted) SecondBaseConfiguration. They apply to more than just logging. Consul registration, metrics appending, etc. The core module can start the logging module in the same way it starts the WebConsole. Maybe I can do that in another commit though.

acidmoose commented on jsonlogging/src/main/java/org/secondbase/logging/JsonLoggerModule.java:
```diff
@@ -0,0 +1,30 @@
+    @Flag(name = "datacenter",
+            description = "The datacenter the service runs in (eu-west-1|ap-southeast-1).")
+    private static String datacenter = "eu-west-1";
+
```
Add another flag to enable JUL handling. Something like:

@Flag(name = "convert-jul-logs")
private static boolean convertJulLogs = false;

neumayer replied:
Yeah, that would currently be possible but the same limitation I mentioned before applies. Where do we call that? Anyway, I think this is something that we should decide at a later point in time in the spirit of not letting this pr getting out of hand.

acidmoose commented on jsonlogging/src/main/java/org/secondbase/logging/JsonLogger.java:
```diff
@@ -0,0 +1,215 @@
+        final LoggerContext loggerContext = (LoggerContext) LoggerFactory.getILoggerFactory();
+        loggerContext.reset();
+
+        if (Strings.isNullOrEmpty(service)) {
```
This is always true, since it has a default value. Unless someone sets it specifically to zero. I think it should not have default values, so that it fails when additional options are needed.

neumayer replied:
I don't think this class and the parameters should have a dependency. I think. I think the check here is reasonable anyway. I agree that they maybe shouldn't have a default value in the module.

neumayer replied:
I updated the pr so non-default value flags are handled. Kind of. This is also the main thing I want to discuss.

acidmoose commented on jsonlogging/src/main/java/org/secondbase/logging/JsonLogger.java:
```diff
@@ -0,0 +1,215 @@
+            throw new IllegalArgumentException("Service name is required");
+        }
+
+        loggerContext.putProperty(DATACENTER, datacenter);
```
Can we put the property here only if it's set? This would include a {datacenter:""} if the datacenter is not set, right?

neumayer replied:
Right now this is how it behaves (i.e. what you think it does not). Null values are not output. This might be just luck, but unless logback changes it's implementation I see no reason to do anything about this right now.

Commit:
Add jul to slf4j bridge in example

Commit:
Use only properties for dependency versions

Commit:
Rename logger classes

Commit:
Make example work with non-default flags

Commit:
Refactor module interface

We add a second method that is invoked after parameters are parsed and
updated.

Commit:
Merge branch 'master' into logging

acidmoose commented on webconsole/src/main/java/org/secondbase/webconsole/HttpWebConsole.java:
```diff
@@ -13,6 +13,8 @@
 import org.secondbase.core.moduleconnection.WebConsole;
 import org.secondbase.webconsole.widget.Widget;
 
+import static org.secondbase.webconsole.WebConsoleConfiguration.port;
```
A rebase missed this one I think.

Commit:
Add line about logging in readme

Commit:
Merge branch 'master' into logging
